### ISSUE TITLE
Karmada-controller-manager stale state

### ISSUE THREAD

**Issue by LavredisG** [karmada-io/karmada#5839]
**Please provide an in-depth description of the question you have**:
I started with a setup of 2 kind clusters, `karmada-host` and `member1`. The former was used both as the host were Karmada was installed, as well as a member cluster. So I had 2 member clusters joined. I also had set kube-prometheus-stack on `karmada-host` in order to retrieve karmada metrics. More specifically I was scraping karmada-controller-manager. While inspecting the `cluster-node-number` metric I was getting the expected result.

However, once I deleted `member1` cluster, karmada-controller-manager returned info about 2 clusters instead of 1 when querried.

**What do you think about this question?**:
Is that expected behavior? Shouldn't karmada-controller-manager observer that 1 cluster has been deleted so that it shouldn't retrieve info about it anymore?

**Environment**:
- Karmada version: built from source 
- Kubernetes version: v1.31.0
- kubectl: v1.29.0

**Comment by RainbowMango** [comment]
> While inspecting the cluster-node-number metric I was getting the expected result.

Firstly, I don't remember having that metric emitted by the karmada-controller-manager. 
Are you sure that's from karmada-controller-manager?

> However, once I deleted member1 cluster,

How do you delete the cluster? By `karmadactl unjoin` or just removed the kind cluster, by a command like `kind delete clusters --name=member1`?

**Comment by LavredisG** [comment]
I am referring to this one:

![image](https://github.com/user-attachments/assets/7ff1149f-1c82-4ffa-be56-ad4adb1b7c6b)

I first unjoined the cluster and then deleted it from kind. A simple workaround is to delete the controller pod, so that when it gets recreated it picks up the correct state, but I don't know if that's the expected way to do it.

**Comment by RainbowMango** [comment]
Oh, I see. That was indeed [reported from Karmada](https://github.com/karmada-io/karmada/blob/adef1e59748e1e1d31cb75fffe406b5dd69a66d7/pkg/metrics/cluster.go#L49-L53). 
We didn't clean up the gauge after the cluster was gone.

> A simple workaround is to delete the controller pod, so that when it gets recreated it picks up the correct state, but I don't know if that's the expected way to do it.

I think it's not the expected behavior. We need more explore how to fix it.

By the way, how important is the metric to you?

**Comment by RainbowMango** [comment]
By the way, let me know if you are going to explore this?

**Comment by LavredisG** [comment]
My goal is to create a custom scheduler for karmada, utilizing resource/power/network metrics, but I have not concluded which I will use, so at this point I was just trying to set up connectivity from karmada metrics to Prometheus.

**Comment by RainbowMango** [comment]
Yeah, I mean if you like you can dig into this issue, find the root cause, then try to fix it :)

**Comment by LavredisG** [comment]
> Yeah, I mean if you like you can dig into this issue, find the root cause, then try to fix it :)

Unfortanutely, I don't think I have enough time to solve it in the foreseeable future :/

**Comment by RainbowMango** [comment]
That's OK. 
Given this issue is part of #5843, let's track it by that issue.
/close
Duplicate of #5843
